lib/cn/kvset.c

```diff
@@ -3633,6 +3674,112 @@ kvset_minkey(struct kvset *ks, const void **minkey, u16 *minklen)
     *minklen = ks->ks_minklen;
 }
 
+struct kvset_vgroup_map *
+kvset_vgmap_alloc(uint32_t nvgroups)
+{
+    struct kvset_vgroup_map *vgmap;
+    size_t sz = sizeof(*vgmap) + 3 * nvgroups * sizeof(*(vgmap->vbidx_out));
+
+    vgmap = calloc(1, sz);
+    if (vgmap) {
+        vgmap->nvgroups = nvgroups;
+        if (nvgroups > 0) {
+            vgmap->vbidx_out = (void *)(vgmap + 1);
+            vgmap->vbidx_adj = vgmap->vbidx_out + nvgroups;
+            vgmap->vbidx_src = vgmap->vbidx_adj + nvgroups;
+        }
+    }
+
+    return vgmap;
+}
+
+void
+kvset_vgmap_free(struct kvset_vgroup_map *vgmap)
+{
+    free(vgmap);
+}
+
+/* TODO: Use binary search as an optimization */
+int
+kvset_vgmap_vbidx_out2adj(struct kvset_vgroup_map *vgmap, int vbidx_out)
+{
+    int i;
+
+    for (i = 0; i < vgmap->nvgroups; i++) {
+        if (vbidx_out <= vgmap->vbidx_out[i])
+            return vgmap->vbidx_adj[i];
+    }
+
+    assert(0);
+    return -1;
+}
+
+/* TODO: Use binary search as an optimization */
+int
+kvset_vgmap_vbidx_src2out(struct kvset_vgroup_map *vgmap, int vbidx_src)
+{
+    int i;
+
+    for (i = 0; i < vgmap->nvgroups; i++) {
+        if (vbidx_src <= vgmap->vbidx_src[i]) {
+            assert(vbidx_src >= vgmap->vbidx_adj[i]);
+            return vbidx_src - vgmap->vbidx_adj[i];
+        }
+    }
+
+    assert(0);
+    return -1;
+}
+
+int
+kvset_vgmap_vbidx_out_start(struct kvset *ks, int vgidx)
+{
+    return vgidx == 0 ? 0 : ks->ks_vgmap->vbidx_out[vgidx - 1] + 1;
+}
+
+int
+kvset_vgmap_vbidx_out_end(struct kvset *ks, int vgidx)
+{
+    return ks->ks_vgmap->vbidx_out[vgidx];
+}
+
+/**
+ * vgmap_src: source vgmap
+ *     - Pass NULL if kblocs are rewritten by the compaction op (k/kv-compact, ingest, spill)
+ *     - Pass source kvset's vgmap if kblocks are not rewritten by the compaction op (kv-split)
+ *
+ * vbidx_src_out: output vblock index of the last vblock in a source kvset's vgroup
+ *
+ * vbidx_tgt: target vgroup map
+ *
+ * vbidx_tgt_out: output vblock index of the last vblock in a target kvset's vgroup
+ *
+ * vbidx_out = vbidx_tgt_out
+ * vbidx_adj = vbidx_adj(vbidx_src_out) in vgmap_src + (vbidx_src_out - vbidx_tgt_out)
+ * vbidx_src = vbidx_out + vbidx_adj
+ */
+void
+kvset_vgmap_vbidx_set(
+    struct kvset_vgroup_map *vgmap_src,
+    uint32_t                 vbidx_src_out,
+    struct kvset_vgroup_map *vgmap_tgt,
+    uint32_t                 vbidx_tgt_out,
+    uint32_t                 vgidx)
+{
+    uint32_t vbidx_src_adj;
+
+    INVARIANT(vgmap_tgt);
+    assert(vgidx < vgmap_tgt->nvgroups);
+
+    vbidx_src_adj = vgmap_src ? kvset_vgmap_vbidx_out2adj(vgmap_src, vbidx_src_out) : 0;
+
+    vgmap_tgt->vbidx_out[vgidx] = vbidx_tgt_out;
+
+    vgmap_tgt->vbidx_adj[vgidx] = vbidx_src_adj + (vbidx_src_out - vbidx_tgt_out);
+
+    vgmap_tgt->vbidx_src[vgidx] = vgmap_tgt->vbidx_out[vgidx] + vgmap_tgt->vbidx_adj[vgidx];
+}
+
```

Comment:
s/kblocs/kblocks/
s/op (/op:/
i.e., by the compaction op:  k/kv-compact, ingest, spill.